tools/azure-sdk-tools/ci_tools/github_tools.py

```diff
@@ -0,0 +1,356 @@
+        checkout_with_fetch(folder, "pull/{}/head".format(pr_number))
+    # If there is SHA1, checkout it. If PR number was given, SHA1 could be inside that PR.
+    if branch_or_commit:
+        repo = Repo(str(folder))
+        repo.git.checkout(branch_or_commit)
+
+
+def do_pr(
+    gh_token, sdk_git_id, sdk_pr_target_repo_id, branch_name, base_branch, pr_body=""
+):  # pylint: disable=too-many-arguments
+    "Do the PR"
+    if not gh_token:
+        _LOGGER.info("Skipping the PR, no token found")
+        return None
+    if not sdk_pr_target_repo_id:
+        _LOGGER.info("Skipping the PR, no target repo id")
+        return None
+
+    github_con = Github(gh_token)
+    sdk_pr_target_repo = github_con.get_repo(sdk_pr_target_repo_id)
+
+    if "/" in sdk_git_id:
+        sdk_git_owner = sdk_git_id.split("/")[0]
+        _LOGGER.info("Do the PR from %s", sdk_git_owner)
+        head_name = "{}:{}".format(sdk_git_owner, branch_name)
+    else:
+        head_name = branch_name
+        sdk_git_repo = github_con.get_repo(sdk_git_id)
+        sdk_git_owner = sdk_git_repo.owner.login
+
+    try:
+        github_pr = sdk_pr_target_repo.create_pull(
+            title="Automatic PR from {}".format(branch_name), body=pr_body, head=head_name, base=base_branch
+        )
+    except GithubException as err:
+        if err.status == 422 and err.data["errors"][0].get("message", "").startswith("A pull request already exists"):
+            matching_pulls = sdk_pr_target_repo.get_pulls(base=base_branch, head=sdk_git_owner + ":" + head_name)
+            matching_pull = matching_pulls[0]
+            _LOGGER.info("PR already exists: %s", matching_pull.html_url)
+            return matching_pull
+        raise
+    _LOGGER.info("Made PR %s", github_pr.html_url)
+    return github_pr
+
+
+def remove_readonly(func, path, _):
+    "Clear the readonly bit and reattempt the removal"
+    os.chmod(path, stat.S_IWRITE)
+    func(path)
+
+
+@contextmanager
+def manage_git_folder(gh_token, temp_dir, git_id, *, pr_number=None):
+    """Context manager to avoid readonly problem while cleanup the temp dir.
+
+    If PR number is given, use magic branches "pull" from Github.
+    """
+    _LOGGER.debug("Git ID %s", git_id)
+    if Path(git_id).exists():
+        yield git_id
+        return  # Do not erase a local folder, just skip here
+
+    # Clone the specific branch
+    split_git_id = git_id.split("@")
+    branch = split_git_id[1] if len(split_git_id) > 1 else None
+    clone_to_path(gh_token, temp_dir, split_git_id[0], branch_or_commit=branch, pr_number=pr_number)
+    try:
+        yield temp_dir
+        # Pre-cleanup for Windows http://bugs.python.org/issue26660
+    finally:
+        _LOGGER.debug("Preclean Rest folder")
+        shutil.rmtree(temp_dir, onerror=remove_readonly)
+
+
+class GithubLink:
+    def __init__(self, gitid, link_type, branch_or_commit, path, token=None):  # pylint: disable=too-many-arguments
```

Comment:
I'm treating these as a bit of a `lift and shift`. Ideally, we would type all this.....but given you didn't really modify this file at all other than location, I feel like a jerk asking for it.

We also _use_ `git_tools` in `update_pr` and that's pretty much it, so I'm not certain it's WORTH IT to type it all.

Reply:
Yep, you're totally right about this just being lifted from its old location. I would also love to type this all, but after looking through a small chunk of this file alone, it would definitely take some time to even reverse engineer the typing since I'm completely unfamiliar with the code. I agree that it's probably not worth it given the investment vs. payoff, so I'll revisit this at a later time maybe